ot/bregman.py

```diff
@@ -1787,3 +1789,287 @@ def empirical_sinkhorn_divergence(X_s, X_t, reg, a=None, b=None, metric='sqeucli
+
+        if len(J) != nt_budget:
+            if uniform:
+                bK = b / K_sum_rows
+                bK_sort = np.sort(bK)[::-1]
+            epsilon_v_square = bK_sort[ns_budget - 1:ns_budget + 1].mean()
+            J = np.where(b >= epsilon_v_square * K_sum_rows)[0].tolist()
+
+        epsilon = (epsilon_u_square * epsilon_v_square) ** (1 / 4)
+        kappa = (epsilon_v_square / epsilon_u_square) ** (1 / 2)
+
+        if verbose:
+            print("Epsilon = %s\n" %epsilon)
+            print("Scaling factor = %s\n" %kappa)
+
+        if verbose:
+            print('|I_active| = %s \t |J_active| = %s ' %(len(I), len(J)))
+
+        # Ic, Jc: complementary of the active sets I and J
+        Ic = list(set(list(range(ns))) - set(I))
+        Jc = list(set(list(range(nt))) - set(J))
+
+        K_IJ = K[np.ix_(I, J)]
+        K_IcJ = K[np.ix_(Ic, J)]
+        K_IJc = K[np.ix_(I, Jc)]
+
+        K_min = K_IJ.min()
+        if K_min == 0:
+            K_min = np.finfo(float).tiny
+
+        # a_I, b_J, a_Ic, b_Jc
+        a_I = a[I]
+        b_J = b[J]
+        if not uniform:
+            a_I_min = a_I.min()
+            a_I_max = a_I.max()
+            b_J_max = b_J.max()
+            b_J_min = b_J.min()
+        else:
+            a_I_min = a_I[0]
+            a_I_max = a_I[0]
+            b_J_max = b_J[0]
+            b_J_min = b_J[0]
+
+        # box constraints in L-BFGS-B (see Proposition 1 in [1])
+        bounds_u = [(max(kappa * a_I_min / (epsilon * (nt - ns_budget) + ns_budget * (b_J_max / (
+                epsilon * ns * K_min))), epsilon / kappa), a_I_max / (epsilon * nt * K_min))] * ns_budget
+
+        bounds_v = [(max(b_J_min / (epsilon * (ns - ns_budget) + ns_budget * (a_I_max / (epsilon * nt * K_min))), \
+                         epsilon * kappa), b_J_max / (epsilon * ns * K_min))] * ns_budget
+
+        # pre-calculated constants for the objective
+        vec_eps_IJc = epsilon * kappa * (K_IJc * np.ones(nt - ns_budget).reshape((1, -1))).sum(axis=1)
+        vec_eps_IcJ = (epsilon / kappa) * (np.ones(ns - ns_budget).reshape((-1, 1)) * K_IcJ).sum(axis=0)
+
+    # initialisation
+    u0 = np.full(ns_budget, (1. / ns_budget) + epsilon / kappa)
+    v0 = np.full(nt_budget, (1. / nt_budget) + epsilon * kappa)
+
+    # pre-calculed constants for Restricted Sinkhorn (see Algorithm 2 in [1])
+    if restricted:
+        if ns_budget != ns or ns_budget != nt:
+            cst_u = kappa * epsilon * K_IJc.sum(axis=1)
+            cst_v = epsilon * K_IcJ.sum(axis=0) / kappa
+
+        cpt = 1
+        while (cpt < 5):  # 5 iterations
+            K_IJ_v = K_IJ.T @ u0 + cst_v
+            v0 = b_J / (kappa * K_IJ_v)
+            KIJ_u = K_IJ @ v0 + cst_u
+            u0 = (kappa * a_I) / KIJ_u
+            cpt += 1
+
+        u0 = projection(u0, epsilon / kappa)
+        v0 = projection(v0, epsilon * kappa)
+
+    else:
+        u0 = u0
+        v0 = v0
+
+    def restricted_sinkhorn(usc, vsc, max_iter=5):
```

Comment:
Naive question: why 5 for max_iter ?

Reply:
It's just a guess value :) to give a good warm-start point from the LBFGS solver.  I think it should be greater than 3 and less than 5 (the more given iterations in restricted_sinkhorn function, the more spending time for the LBFGS initialization)